src/Data/Parameterized/MonoidSolver.hs

```diff
@@ -0,0 +1,210 @@
+  Normalize me = UnDiff (Diff me)
+
+-- | Note on termination: The one recursive call to this function is strictly
+-- decreasing on its first argument. This function is mutually recursive with
+-- 'normalizeSound' and recursive calls to that function are made on the second
+-- argument, but calls to this function *from* 'normalizeSound' are on strict
+-- subexpressions of its argument.
+normalizeLemma ::
+  forall k f me1 me2.
+  (forall proxy n. proxy n -> f (Op k (Unit k) n) :~: f n) ->
+  (forall proxy n. proxy n -> f (Op k n (Unit k)) :~: f n) ->
+  (forall proxy n m l. proxy n -> proxy m -> proxy l -> f (Op k n (Op k m l)) :~: f (Op k (Op k n m) l)) ->
+  MERepr k f me1 ->
+  MERepr k f me2 ->
+  f (Eval k (Apply (Diff me1) (Apply (Diff me2) 'EUnit))) :~:
+    f (Op k (Eval k me1) (Eval k me2))
+normalizeLemma idl idr assoc mer1 mer2 =
+  case mer1 of
+    EUnitRepr ->
+      case idl (Proxy :: Proxy (Eval k me2)) of
+        Refl ->
+          case norm mer2 of
+            Refl -> Refl
+    EVarRepr {} ->
+      case norm mer2 of
+        Refl -> Refl
+    EOpRepr (mer1' :: MERepr k f me1') (mer2' :: MERepr k f me2') ->
+      case normalizeLemma idl idr assoc mer1' (EOpRepr mer2' mer2) of
+        Refl ->
+              assoc (Proxy :: Proxy (Eval k me1')) (Proxy :: Proxy (Eval k me2')) (Proxy :: Proxy (Eval k me2))
+  where
+    norm :: MERepr k f me -> f (Eval k (Normalize me)) :~: f (Eval k me)
+    norm = normalizeSound idl idr assoc
+
+normalizeSound ::
+  (forall proxy n. proxy n -> f (Op k (Unit k) n) :~: f n) ->
+  (forall proxy n. proxy n -> f (Op k n (Unit k)) :~: f n) ->
+  (forall proxy n m l. proxy n -> proxy m -> proxy l -> f (Op k n (Op k m l)) :~: f (Op k (Op k n m) l)) ->
+  MERepr k f me ->
+  f (Eval k (Normalize me)) :~: f (Eval k me)
+normalizeSound idl idr assoc =
+  \case
+    EUnitRepr -> Refl
+    EVarRepr (sing :: f n) -> idr sing
+    EOpRepr (mer1 :: MERepr k f me1) (mer2 :: MERepr k f me2) ->
+      normalizeLemma idl idr assoc mer1 mer2
+
+solve ::
+  (forall proxy n. proxy n -> f (Op k (Unit k) n) :~: f n) ->
+  (forall proxy n. proxy n -> f (Op k n (Unit k)) :~: f n) ->
+  (forall proxy n m l. proxy n -> proxy m -> proxy l -> f (Op k n (Op k m l)) :~: f (Op k (Op k n m) l)) ->
+  MERepr k f me1 ->
+  MERepr k f me2 ->
+  f (Eval k (Normalize me1)) :~: f (Eval k (Normalize me2)) ->
+  f (Eval k me1) :~: f (Eval k me2)
+solve idl idr assoc repr1 repr2 Refl =
+  case (normalizeSound idl idr assoc repr1, normalizeSound idl idr assoc repr2) of
+    (Refl, Refl) -> Refl
+
+solveNat ::
+  MERepr Nat proxy me1 ->
+  MERepr Nat proxy me2 ->
+  proxy (Eval Nat (Normalize me1)) :~: proxy (Eval Nat (Normalize me2)) ->
+  proxy (Eval Nat me1) :~: proxy (Eval Nat me2)
+solveNat =
+  solve
+    (const Refl)
+    (const Refl)
+    (\proxy1 proxy2 proxy3 ->
+       case plusAssoc proxy1 proxy2 proxy3 of
+         Refl -> Refl)
+
+----------------------------------------------------------------------
+-- Examples
```

Comment:
These examples would be a better fit in the test directory, I think.

Reply:
Agreed, this is part of why this is still WIP :-)